pyccel/parser/syntactic.py

```diff
@@ -535,43 +505,33 @@ def _visit_UnaryOp(self, stmt):
                           bounding_box=(stmt.lineno, stmt.col_offset),
                           severity='fatal')
 
-        if isinstance(target,  PyccelOperator) and target.p  <= Func.p:
-            target = PyccelAssociativeParenthesis(target)
-
-        return PyccelUnary(Func(target))
+        return Func(PyccelUnary(target))
 
     def _visit_BinOp(self, stmt):
 
         first  = self._visit(stmt.left)
         second = self._visit(stmt.right)
 
         if isinstance(stmt.op, ast.Add):
-            expr = PyccelAdd(first, second)
-            return change_priority(expr)
+            return PyccelAdd(first, second)
 
         elif isinstance(stmt.op, ast.Mult):
-            expr = PyccelMul(first, second)
-            return change_priority(expr)
+            return PyccelMul(first, second)
```

Comment:
there is a problem here, the function take in consideration only the first and the second condition.
for example this python code 

```
def f(x, y):
    while x==10 or y==10 or x == 20:
        x = x + 1
    return 0
```

converted to this : 

in fortran 

```
....

function f(x, y) result(Dummy_266)

  implicit none
  integer(kind=8)  :: Dummy_266  
  integer(kind=8), value  :: x
  integer(kind=8), value  :: y

  do while (x == 10_8  .or. y == 10_8 ) 
    x = x + 1_8
  end do
  Dummy_266 = 0_8
  return
end function

....
```
in C 

```
long f(long x, long y)
{
    long Dummy_102;
    while (x == 10 || y == 10)
    {
        x = x + 1;
    }
    Dummy_102 = 0;
    return Dummy_102;
}

```

we've made a a solution in the [bug_parentheses](https://github.com/pyccel/pyccel/tree/bug_parentheses) branch

Reply:
Good catch! You are right, unlike binary operators, boolean operators can take any number of arguments.

This is actually a different bug to the one reported in issue 392. If you find something like this please open a new issue. If you plan to fix it yourself you can assign yourself the issue so everyone else knows it's being looked at.

I have used your solution to fix this.